### PR TITLE
Replace align_cfg::gap with align_cfg::gap_cost_affine

### PR DESCRIPTION
**Pull request by wvandertoorn**
Fixes https://github.com/seqan/product_backlog/issues/176.

Replaces seqan3::align_cfg::gap by seqan3::align_cfg::gap_cost_affine.
Deprecates seqan3::gap_scheme,seqan3::gap_open_score and  seqan3::gap_score.

### CODE REVIEW

File: include/seqan3/alignment/configuration/align_config_gap_cost_affine.hpp
Comment (simonsasse) [inline include/seqan3/alignment/configuration/align_config_gap_cost_affine.hpp:28]: I'm not sure, but I think you missed a ` here :)
```suggestion
/*!\brief A strong type of underlying type `int32_t` that represents a score (usually negative) that
```

File: include/seqan3/alignment/configuration/align_config_gap_cost_affine.hpp
Comment (simonsasse) [inline include/seqan3/alignment/configuration/align_config_gap_cost_affine.hpp:44]: And also here:
```suggestion
 /*!\brief A strong type of underlying type `int32_t` that represents the score (usually negative) of any character
```

File: doc/tutorial/pairwise_alignment/pairwise_alignment_solution_3.cpp
Comment (simonsasse) [inline doc/tutorial/pairwise_alignment/pairwise_alignment_solution_3.cpp:2]: I was told to instead of include <xxx.all> include single files. This speeds up compiletime. I don't know if this also applies to documentation and so on. Maybe wait for the second reviewer with that :)

File: test/unit/alignment/configuration/align_config_gap_cost_affine_test.cpp
Comment (simonsasse) [inline test/unit/alignment/configuration/align_config_gap_cost_affine_test.cpp:13]: Travis throws errors because it doesn't know the get function I think. So this should fix it I hope. ( https://docs.seqan.de/seqan/3-master-user/group__algorithm.html#ga8f02011be28ea1c73ca39a46fd5ce613 )
```suggestion
 using seqan3::get; 
```

File: doc/tutorial/pairwise_alignment/configurations.cpp
Comment (simonsasse) [inline doc/tutorial/pairwise_alignment/configurations.cpp:69]: Maybe this should also be changed ?
```suggestion
//! [gap_cost_affine]
```

File: doc/tutorial/pairwise_alignment/configurations.cpp
Comment (simonsasse) [inline doc/tutorial/pairwise_alignment/configurations.cpp:77]:
```suggestion
//! [gap_cost_affine]
```

File: test/snippet/alignment/configuration/align_cfg_gap_cost_affine_example.cpp
Comment (simonsasse) [inline test/snippet/alignment/configuration/align_cfg_gap_cost_affine_example.cpp:5]: This might be a bit more exact :)
```suggestion
    // Configuration with linear gap costs.
```

File: test/snippet/alignment/configuration/align_cfg_gap_cost_affine_example.cpp
Comment (simonsasse) [inline test/snippet/alignment/configuration/align_cfg_gap_cost_affine_example.cpp:9]:
```suggestion
    // Configuration with affine gap costs. Score for opening a gap during the alignment algorithm will be -11.
```

File: test/unit/alignment/pairwise/fixture/semi_global_affine_banded.hpp
Comment (simonsasse) [inline test/unit/alignment/pairwise/fixture/semi_global_affine_banded.hpp:29]:
```suggestion

```

File: test/unit/alignment/pairwise/fixture/semi_global_affine_unbanded.hpp
Comment (simonsasse) [inline test/unit/alignment/pairwise/fixture/semi_global_affine_unbanded.hpp:28]:
```suggestion

```

**Review by simonsasse — COMMENTED** [review]
Sorry that I commented before being done with the review. Here are some more tiny things 🧐😊

File: doc/tutorial/pairwise_alignment/pairwise_alignment_solution_3.cpp
Comment (wvandertoorn) [reply doc/tutorial/pairwise_alignment/pairwise_alignment_solution_3.cpp:2]: You're right for the library itself, but I believe that for the tutorials the `xxx/all.hpp` includes are preferred.

**Review by simonsasse — APPROVED** [review]
Looks good, thanks 👍

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp
Comment (marehr) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp:328]:
```suggestion
            align_cfg::gap_cost_affine edit_gap_cost{};
            auto const & gap_cost = config_with_result_type.get_or(edit_gap_cost);
```

File: include/seqan3/alignment/pairwise/policy/all.hpp
Comment (marehr) [inline include/seqan3/alignment/pairwise/policy/all.hpp:76]: How is this related to this PR?

File: test/snippet/alignment/configuration/align_cfg_gap_cost_affine_example.cpp
Comment (marehr) [inline test/snippet/alignment/configuration/align_cfg_gap_cost_affine_example.cpp:11]: Can you add an example how to access the member>

File: test/snippet/core/algorithm/configuration_get.cpp
Comment (marehr) [inline test/snippet/core/algorithm/configuration_get.cpp:18]:
```suggestion
    seqan3::debug_stream << get<1>(my_cfg).lower_diagonal << '\n';  // prints -4
    seqan3::debug_stream << get<seqan3::align_cfg::band_fixed_size>(my_cfg).upper_diagonal << '\n';  // prints 4
```

T_T those whitespaces

File: test/unit/alignment/configuration/align_config_edit_test.cpp
Comment (marehr) [inline test/unit/alignment/configuration/align_config_edit_test.cpp:14]:
```suggestion
```

move this down

File: test/unit/alignment/configuration/align_config_edit_test.cpp
Comment (marehr) [inline test/unit/alignment/configuration/align_config_edit_test.cpp:34]:
```suggestion
{
    using seqan3::get;

```

Comment (marehr) [reply test/unit/alignment/configuration/align_config_edit_test.cpp:34]: Why did you remove `seqan3::get` ?

File: test/unit/alignment/configuration/align_config_edit_test.cpp
Comment (marehr) [inline test/unit/alignment/configuration/align_config_edit_test.cpp:35]:
```suggestion
    auto gap_cost_scheme = seqan3::get<seqan3::align_cfg::gap_cost_affine>(seqan3::align_cfg::edit_scheme);
```

Is this not working?

File: test/unit/alignment/configuration/align_config_gap_cost_affine_test.cpp
Comment (marehr) [inline test/unit/alignment/configuration/align_config_gap_cost_affine_test.cpp:14]: If this is really needed please move this into the tests.

**Review by marehr — CHANGES_REQUESTED** [review]
Thank you :) My first pass-through. I asked the others if we really want to make the `affine.extension_score` member a strong type.

File: test/unit/alignment/configuration/align_config_edit_test.cpp
Comment (wvandertoorn) [reply test/unit/alignment/configuration/align_config_edit_test.cpp:32]: To quote @rrahn :

This is a bug in the implementation of the configuration. So you need to call it unqualified, i.e. `get<seqan3::align_cfg::gap_cost_affine>(cfg)`.
The reason for this is, that the configuration class inherits from `std::tuple` and also it's get interface. But we added one additional interface to call it with template types. This was ok before, since `seqan3::align_cfg::gap` was a class template. But now this changed and `seqan3::align_cfg::gap_cost_affine` is a regular class without templates. In this case it was not finding a suitable candidate, because it had to use `std::get` instead.

File: test/unit/alignment/configuration/align_config_edit_test.cpp
Comment (wvandertoorn) [reply test/unit/alignment/configuration/align_config_edit_test.cpp:35]: no

File: include/seqan3/alignment/pairwise/policy/all.hpp
Comment (wvandertoorn) [reply include/seqan3/alignment/pairwise/policy/all.hpp]: It is not. I forgot to split it out into a separate PR.

File: test/unit/alignment/configuration/align_config_edit_test.cpp
Comment (marehr) [reply test/unit/alignment/configuration/align_config_edit_test.cpp:35]: Thank you :)

**Review by marehr — COMMENTED** [review]
Thank you!

This is basically an approval, can you rebase and then reassign me?

There is one open discussion in the core team, whether 

```c++
    // Accessing the members of the gap scheme
    int open = affine_cfg.open_score.get(); // open == -1
    int extension = affine_cfg.extension_score.get(); // extension == -10
```

or

```c++
    // Accessing the members of the gap scheme
    int open = affine_cfg.open_score; // open == -1
    int extension = affine_cfg.extension_score; // extension == -10
```

So that the members of the config element is a strong_type or the bare type.

**Comment by marehr** [comment]
The members of `affine_cfg` shouldn't be strong types.

```c++
    // Accessing the members of the gap scheme
    int open = affine_cfg.open_score; // open == -1
    int extension = affine_cfg.extension_score; // extension == -10
```

We decided on this, see https://github.com/seqan/product_backlog/issues/202

@wvandertoorn Is this enough for you and do you know what this means and can you make the changes? If not I can explain it to you :)

File: include/seqan3/alignment/configuration/align_config_gap_cost_affine.hpp
Comment (marehr) [inline include/seqan3/alignment/configuration/align_config_gap_cost_affine.hpp:79]:
```suggestion
    int32_t open_score{0};
    //!\brief The cost per gap character. Defaults to -1.
    int32_t extension_score{-1};
```

I meant to change it like this :)

File: doc/tutorial/pairwise_alignment/configurations.cpp
Comment (marehr) [inline doc/tutorial/pairwise_alignment/configurations.cpp:79]:
```suggestion
int open_score = affine_scheme.open_score; // == -10
int extension_score = affine_scheme.extension_score; // == -1
//! [gap_cost_affine]
(void) open_score;
(void) extension_score;
```

File: include/seqan3/alignment/scoring/gap_scheme.hpp
Comment (marehr) [inline include/seqan3/alignment/scoring/gap_scheme.hpp:30]:
```suggestion
 *        character. 
 * \deprecated This type is deprecated and will be removed in SeqAn3.1.
```

File: include/seqan3/alignment/scoring/gap_scheme.hpp
Comment (marehr) [inline include/seqan3/alignment/scoring/gap_scheme.hpp:57]:
```suggestion
 *        is incurred once additionaly per stretch of consecutive gaps.T
 * \deprecated This type is deprecated and will be removed in SeqAn3.1.
```

File: test/snippet/core/algorithm/configuration_get.cpp
Comment (marehr) [inline test/snippet/core/algorithm/configuration_get.cpp:15]: :+1: nice catch

**Review by marehr — COMMENTED** [review]
Thank you so much :) (Sorry for the delay :crying_cat_face:)

**Review by marehr — APPROVED** [review]

**Pull request merged**
